src/Data/List/Extra.hs

```diff
@@ -98,6 +99,25 @@ allSame [] = True
 allSame (x:xs) = all (x ==) xs
 
 
+-- | A total 'head' with a default value.
+--
+-- > headDef 1 []      == 1
+-- > headDef 1 [2,3,4] == 2
+-- > \x xs -> headDef x xs == fromMaybe x (listToMaybe xs)
+headDef :: a -> [a] -> a
+headDef d [] = d
+headDef _ (x:_) = x
+
+
+-- | A total 'last' with a default value.
+--
+-- > lastDef 1 []      == 1
+-- > lastDef 1 [2,3,4] == 4
+-- > \x xs -> lastDef x xs == last (x:xs)
+lastDef :: a -> [a] -> a
+lastDef d xs = foldl (\_ x -> x) d xs
```

Comment:
I know this looks weird, but apparently this is the fastest way to do this: https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.List.html#last